File: tests/typescript/tests-typescript-playwright/src/pages/header.page.ts
import { expect, type Locator, type Page } from '@playwright/test';

export class HeaderPage {
  readonly page: Page;
  readonly root: Locator;
  readonly burger: Locator;
  readonly menu: Locator;
  readonly langToggle: Locator;
  readonly langLabel: Locator;
  readonly themeToggle: Locator;
  readonly html: Locator;

  constructor(page: Page) {
    this.page = page;
    this.root = page.getByTestId('header');
    this.burger = page.getByTestId('header-burger');
    this.menu = page.getByTestId('header-menu');
    const tools = page.getByTestId('header-tools');
    this.langToggle = tools.getByTestId('header-lang-toggle');
    this.langLabel = tools.getByTestId('header-lang-label');
    this.themeToggle = tools.getByTestId('header-theme-toggle');
    this.html = page.locator('html');
  }

  activeNav(testid: string): Locator {
    return this.page.getByTestId(testid);
  }

  currentPageLinks(): Locator {
    return this.page.locator("[data-testid='header-nav'] a[aria-current='page']");
  }

  menuNav(testid: string): Locator {
    return this.page.getByTestId(testid);
  }

  async shouldHaveActiveNav(navTestid: string): Promise<void> {
    const item = this.activeNav(navTestid);
    await expect(item).toBeVisible();
    await expect(item).toHaveClass(/is-active/);
    await expect(item).toHaveAttribute('aria-current', 'page');
    await expect(this.currentPageLinks()).toHaveCount(1);
  }

  async shouldHaveActiveMenuNav(menuNavTestid: string): Promise<void> {
    const item = this.menuNav(menuNavTestid);
    await expect(item).toBeVisible();
    await expect(item).toHaveClass(/is-active/);
    await expect(item).toHaveAttribute('aria-current', 'page');
  }

  async clickNav(testid: string): Promise<void> {
    await this.activeNav(testid).click();
  }

  async setMobileViewport(): Promise<void> {
    await this.page.setViewportSize({ width: 375, height: 812 });
  }

  async resetViewport(): Promise<void> {
    await this.page.setViewportSize({ width: 1280, height: 720 });
  }

  async openMenu(): Promise<void> {
    await this.burger.click();
    await this.menu.waitFor({ state: 'visible' });
  }

  async clickMenuNav(testid: string): Promise<void> {
    await this.menuNav(testid).click();
  }

  async shouldHaveClosedMenu(): Promise<void> {
    await this.menu.waitFor({ state: 'hidden' });
    await expect(this.burger).toHaveAttribute('aria-expanded', 'false');
  }

  async clickLangToggle(): Promise<void> {
    await this.langToggle.click();
  }

  async clickThemeToggle(): Promise<void> {
    await this.themeToggle.click();
  }
}

File: tests/java/tests-java-junit5-api_request-playwright/src/test/java/pages/HeaderPage.java
package pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.WaitForSelectorState;
import helpers.ViewportHelper;
import io.qameta.allure.Step;

import java.util.regex.Pattern;

import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;

public class HeaderPage {

    private static final Pattern IS_ACTIVE = Pattern.compile("is-active");

    private final Page page;
    public final Locator root;
    public final Locator burger;
    public final Locator menu;
    public final Locator langToggle;
    public final Locator langLabel;
    public final Locator themeToggle;
    public final Locator html;

    public HeaderPage(Page page) {
        this.page = page;
        this.root = page.getByTestId("header");
        this.burger = page.getByTestId("header-burger");
        this.menu = page.getByTestId("header-menu");
        var tools = page.getByTestId("header-tools");
        this.langToggle = tools.getByTestId("header-lang-toggle");
        this.langLabel = tools.getByTestId("header-lang-label");
        this.themeToggle = tools.getByTestId("header-theme-toggle");
        this.html = page.locator("html");
    }

    public Locator activeNav(String testid) {
        return page.getByTestId(testid);
    }

    public Locator currentPageLinks() {
        return page.locator("[data-testid='header-nav'] a[aria-current='page']");
    }

    public Locator menuNav(String testid) {
        return page.getByTestId(testid);
    }

    @Step("Desktop nav '{navTestid}' is the active item")
    public HeaderPage shouldHaveActiveNav(String navTestid) {
        var item = activeNav(navTestid);
        assertThat(item).isVisible();
        assertThat(item).hasClass(IS_ACTIVE);
        assertThat(item).hasAttribute("aria-current", "page");
        assertThat(currentPageLinks()).hasCount(1);
        return this;
    }

    @Step("Menu nav '{menuNavTestid}' is the active item")
    public HeaderPage shouldHaveActiveMenuNav(String menuNavTestid) {
        var item = menuNav(menuNavTestid);
        assertThat(item).isVisible();
        assertThat(item).hasClass(IS_ACTIVE);
        assertThat(item).hasAttribute("aria-current", "page");
        return this;
    }

    @Step("Click header nav {testid}")
    public HeaderPage clickNav(String testid) {
        activeNav(testid).click();
        return this;
    }

    @Step("Emulate mobile viewport (375x812)")
    public HeaderPage setMobileViewport() {
        ViewportHelper.setViewport(375, 812);
        return this;
    }

    @Step("Reset viewport to default")
    public HeaderPage resetViewport() {
        ViewportHelper.resetViewport();
        return this;
    }

    @Step("Open the burger menu")
    public HeaderPage openMenu() {
        burger.click();
        menu.waitFor();
        return this;
    }

    @Step("Click menu nav link {testid}")
    public HeaderPage clickMenuNav(String testid) {
        menuNav(testid).click();
        return this;
    }

    @Step("Wait until the burger menu is closed")
    public HeaderPage shouldHaveClosedMenu() {
        menu.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.HIDDEN));
        return this;
    }

    @Step("Click language toggle")
    public HeaderPage clickLangToggle() {
        langToggle.click();
        return this;
    }

    @Step("Click theme toggle")
    public HeaderPage clickThemeToggle() {
        themeToggle.click();
        return this;
    }
}
